src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,545 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Transactions;
+using NHibernate.AdoNet;
+using NHibernate.Engine;
+using NHibernate.Engine.Transaction;
+using NHibernate.Impl;
+using NHibernate.Util;
+
+namespace NHibernate.Transaction
+{
+	/// <summary>
+	/// <see cref="ITransaction"/> factory implementation supporting system
+	/// <see cref="System.Transactions.Transaction"/>.
+	/// </summary>
+	public class AdoNetWithSystemTransactionFactory : AdoNetTransactionFactory
+	{
+		private static readonly IInternalLogger _logger = LoggerProvider.LoggerFor(typeof(ITransactionFactory));
+
+		/// <summary>
+		/// See <see cref="Cfg.Environment.SystemTransactionCompletionLockTimeout"/>.
+		/// </summary>
+		protected int SystemTransactionCompletionLockTimeout { get; private set; }
+		/// <summary>
+		/// See <see cref="Cfg.Environment.UseConnectionOnSystemTransactionPrepare"/>.
+		/// </summary>
+		protected bool UseConnectionOnSystemTransactionPrepare { get; private set; }
+
+		/// <inheritdoc />
+		public override void Configure(IDictionary<string, string> props)
+		{
+			base.Configure(props);
+			SystemTransactionCompletionLockTimeout =
+				PropertiesHelper.GetInt32(Cfg.Environment.SystemTransactionCompletionLockTimeout, props, 5000);
+			if (SystemTransactionCompletionLockTimeout < -1)
+				throw new HibernateException(
+					$"Invalid {Cfg.Environment.SystemTransactionCompletionLockTimeout} value: {SystemTransactionCompletionLockTimeout}. It can not be less than -1.");
+			UseConnectionOnSystemTransactionPrepare =
+				PropertiesHelper.GetBoolean(Cfg.Environment.UseConnectionOnSystemTransactionPrepare, props, true);
+		}
+
+		/// <inheritdoc />
+		public override void EnlistInSystemTransactionIfNeeded(ISessionImplementor session)
+		{
+			// Handle the transaction on the originating session only.
+			var originatingSession = session.ConnectionManager.Session;
+
+			if (!originatingSession.ConnectionManager.ShouldAutoJoinTransaction)
+			{
+				return;
+			}
+
+			// Ensure the session does not run on a thread supposed to be blocked, waiting
+			// for transaction completion.
+			originatingSession.TransactionContext?.Wait();
+
+			var transaction = System.Transactions.Transaction.Current;
+			// We may have defined the transaction context before having the connection, so we
+			// need to ensure enlistment even when the transaction context is already defined.
+			// But avoid redefining to a sub-scope transaction.
+			if (originatingSession.TransactionContext == null ||
+				((SystemTransactionContext)originatingSession.TransactionContext).AmbientTransaction == transaction)
+			{
+				originatingSession.ConnectionManager.EnlistIfRequired(transaction);
+			}
```

Comment:
I think I am going to change that for being able to only have `if (originatingSession.TransactionContext == null)`. The connection manager would store the transaction (does already that), and additionally when acquiring a new connection, enlist it in that stored transaction (instead of checking `Current`: after all, that is the job of the transaction factory to tell in which transaction we have to enlist.

If we happen to be in the invalid case I was writing about in my [previous comment](https://github.com/nhibernate/nhibernate-core/pull/627#discussion_r126184961), that would fail at that point. And that would be fine, as it would avoid having a session enlisted in a scope with its connection in another scope. In other cases, sub-scopes will still be silently ignored. Throwing on them could be an additional option.